Initial version of JanusGraph.Net

FlorianHockmann opened the pull request:
This adds the initial version of the library.

### Included Functionality

* Text predicates
* Edge ids (needed to deserialize edges)
* `Geoshape.point()`

### Still Missing:

* Geo predicates
* Geoshapes beyond point

### Decision Decisions and Dependencies

* Builds are started and configured via [Cake](https://cakebuild.net/), _a cross platform build automation system with a C# DSL_. This enables consistent builds across platforms and also Travis which are very easy to configure. This is also explained in `BUILDING.md`.
* Testing is done via [testcontainers-dotnet](https://github.com/testcontainers/testcontainers-dotnet). The .NET version of Testcontainers is pretty new but it makes integration tests much easier and its Java version is widely used. It automatically starts a Docker container in the background to be used in tests. (@farodin91 is working on integrating it for JanusGraph itself janusgraph/janusgraph#1014.)
* The tests currently use a [Docker image from my own repo](https://hub.docker.com/r/florianhockmann/janusgraph-testing/). That is definitely not a permanent solution. I'd suggest that we publish something like nightly builds as Docker images for those purposes under a JanusGraph repo on Docker Hub.

The PR also includes a commit by @farodin91 who provided improved Cake bootstrapper scripts to execute Cake also on .NET Core which removed the dependency on Mono. (It also reduced the build time from 12 minutes down to under 3 on Travis.)

The PR only includes very limited documentation for the library itself in the `README.md` as we haven't decided yet where docs for these libraries like JanusGraph.Net will be hosted.

mbrukman commented on README.md:
```diff
@@ -1,8 +1,21 @@
+# JanusGraph.Net
 
-## License
+JanusGraph.Net is the .NET driver of [JanusGraph](http://janusgraph.org/). It is based on Apache TinkerPop™'s [Gremlin.Net](http://tinkerpop.apache.org/docs/current/reference/#gremlin-DotNet).
```
If this is "based on Apache TinkerPop's Gremlin.Net`, where is all the code that should be marked as (c) Apache Software Foundation or similar?

All the files in this PR appear to be marked as "(c) JanusGraph.Net authors" of which there are currently only 2 (listed in the `CONTRIBUTORS.txt` file). Note that if any files were copied from Apache TinkerPop, even if they were modified afterwards, must retain the original copyright header. We can then add "portions copyright by JanusGraph.Net authors" if that's appropriate, but we should not be removing existing copyright headers in any imported files.

mbrukman replied:
Also, given that this is based on a different piece of software, this PR should also update `LICENSE.txt` to be more similar to the JanusGraph `LICENSE.txt` in that it talks about "all changes prior to date <X> are under the Apache 2.0 license, and all changes beginning with date <X> are under the dual license of Apache 2.0 for code + CC-BY-4.0 for documentation".

See https://github.com/JanusGraph/janusgraph/blob/master/LICENSE.txt for details.

FlorianHockmann replied:
What I meant with that sentence is that Gremlin.Net is its core library, not that any code is included directly from Gremlin.Net or TinkerPop in general.
JanusGraph.Net just extends Gremlin.Net with support for types that are JanusGraph specific but the core functionality of the driver is provided by Gremlin.Net. Users should now that fact so that they don't think that they have to decide between Gremlin.Net and JanusGraph.Net as the latter is basically just an extension of Gremlin.Net.
So, I think that the license headers and the `LICENSE.txt` are appropriate.

Do you think that the sentence is misleading?

FlorianHockmann commented on .gitattributes:
```diff
@@ -0,0 +1,2 @@
+* text eol=lf
+* text eol=lf
```
I'm sure that the first search result I found on how to enforce `lf` line endings with git showed this pattern and explained it as something like the first is used for check-out and the second for check-in. But I just searched again and couldn't find a source for this. So, I will remove the duplicate line.

FlorianHockmann commented on README.md:
```diff
@@ -1,8 +1,21 @@
-# JanusGraph .NET driver
+# JanusGraph.Net
 
-## License
```
We don't have it in JanusGraph/janusgraph and I wanted to have a short `README.md` with only the most important information, but I can revert this change when you want to keep the `License` section there.

mbrukman replied:
That's a good point; the JanusGraph repo does not have a "License" section, as you described.

Given that JanusGraph license is not as simple as others (so it's not automatically parsed by GitHub), I think it's good to include this, but you can put it as the last thing on the readme after all the other content.

I think we should include a similar paragraph in the main JanusGraph repo as well.

FlorianHockmann commented:
> I am not a .NET expert; please nominate either an existing JanusGraph committer or someone else from the Gremlin.Net community to review this PR for the .NET aspects.

Do we have any @JanusGraph/committers with at least some experience in .NET / C#?
The other Gremlin.Net contributors basically all work for companies with their own graph database. So, I'm not sure whether they want to / can help us with reviews, but that also wouldn't be a good long term solution in my opinion.

FlorianHockmann commented on test/JanusGraph.Net.IntegrationTest/JanusGraph.Net.IntegrationTest.csproj:
```diff
@@ -0,0 +1,17 @@
+    <PackageReference Include="TestContainers" Version="0.0.2.55" />
+    <PackageReference Include="xunit" Version="2.3.1" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
+    <DotNetCliToolReference Include="dotnet-xunit" Version="2.3.1" />
```
They are all open source. Here are the repos and the licenses:
* Microsoft.NET.Test.Sdk: [Repo](https://github.com/Microsoft/vstest) | [MIT License](https://github.com/Microsoft/vstest/blob/master/LICENSE)
* Polly: [Repo](https://github.com/App-vNext/Polly) | [New BSD License](https://github.com/App-vNext/Polly/blob/master/LICENSE.txt)
* TestContainers: [Repo](https://github.com/testcontainers/testcontainers-dotnet) | [MIT License](https://github.com/testcontainers/testcontainers-dotnet/blob/master/LICENSE)
* 3 * xunit: [Repo](https://github.com/xunit/xunit) | [Apache 2 License](https://github.com/xunit/xunit/blob/master/license.txt)

FlorianHockmann replied:
I had to add another dependency to this project to move the config options into a text file:
Microsoft.Extensions.Configuration.Json: [Repo](https://github.com/aspnet/Configuration) | [Apache 2 License](https://github.com/aspnet/Configuration/blob/master/LICENSE.txt)

FlorianHockmann commented on build.sh:
```diff
@@ -0,0 +1,88 @@
+
+    # Download the Cake.CoreCLR nuget package and save it as a *.zip file.
+    if [ ! -f "$ZIP_NAME" ]; then
+        wget -O $ZIP_NAME https://www.nuget.org/api/v2/package/Cake.CoreCLR/$CAKE_VERSION
```
Isn't `wget` the default tool in most Linux distributions? We could however also check which of these tools is installed and then use that.

FlorianHockmann commented on AUTHORS.txt:
```diff
@@ -1,4 +1,4 @@
-# This is the official list of JanusGraph authors for copyright purposes.
+# This is the official list of JanusGraph.Net authors for copyright purposes.
```
It only has 78 chars. Looks like GitHub wraps lines already for something between 75 and 78 chars.

mbrukman replied:
Thanks for clarifying!

FlorianHockmann commented:
Thanks for your reviews so far, @mbrukman and @pluradj! I just pushed a commit to address your comments.

mbrukman commented:
Sorry for the pedantic nature of the CLA verifier, but it really is just simple rather than pedantic (i.e., it simply does an equality comparison on names, GitHub usernames, and emails, without considering any permutations, reorderings, or any other formatting changes).

[One of the commits](https://github.com/JanusGraph/janusgraph-dotnet/pull/1.patch) in this PR has the line:

```
From: "Jansen, Jan" <Jan.Jansen@gdata.de>
```

whereas the name on the CLA is simply "Jan Jansen". Fixing that name should address the CLA issue.

mbrukman commented on .travis.yml:
```diff
@@ -0,0 +1,11 @@
+language: csharp
```
Looks like this was done, thanks!

mbrukman commented on test/JanusGraph.Net.IntegrationTest/GremlinServerContainer.cs:
```diff
@@ -0,0 +1,39 @@
+﻿using System;
```
Please add a license header to this file.

FlorianHockmann replied:
Done, I must have missed this file somehow, but I just searched with PowerShell through all `.cs` files to make sure that there are none left now.

FlorianHockmann commented:
> One of the commits in this PR has the line:
> From: "Jansen, Jan" <Jan.Jansen@gdata.de>
> whereas the name on the CLA is simply "Jan Jansen". Fixing that name should address the CLA issue.

I just changed the author of the commit (as described [here](https://stackoverflow.com/a/28845565/6753576)) and now the CLA bot seems to be happy.

pluradj commented on src/JanusGraph.Net/Geoshapes/Point.cs:
```diff
@@ -0,0 +1,79 @@
+        /// <inheritdoc />
+        public override string ToString()
+        {
+            return $"POINT ({Longitude} {Latitude})";
```
add a comment describing why the order is this way

farodin91 replied:
I think this would be cleaner with directions for example `POINT (10.12345 N -12.3434 E)

FlorianHockmann replied:
@pluradj: What is the reason for the ordering? I only copied it from JanusGraph and only found this sentence in the docs:
> Note that unlike the JanusGraph API GeoJSON specifies coordinates as lng lat.

@farodin91: I only know it in this form if the orientation is added: `51° 28′ 55″ N, 7° 12′ 57″ E`. The current implementation corresponds to the Java implementation. So, we stay consistent across languages.

porunov replied:
I am not aware of specific implementation, but it seems that there is a bug in JanusGraph coordinates. I see that in some places we use `lat lng` and in other places `lng lat`. As a result, there is a confusion which leads to new bugs. Right now I've found a bug in ElasticSearch index implementation https://github.com/JanusGraph/janusgraph/issues/1291 . I didn't check Solr and Lucene but there is a possibility that this bug is presented there also.

P.S. I've never seen `lng lat` order before in other systems. I think that we should stick to one style (I prefer `lat lng`).

FlorianHockmann replied:
JanusGraph expects the form lat lng, but uses [spatial4j](https://github.com/locationtech/spatial4j) internally which seems to use lng lat. I mirrored the Java API here to stay at least consistent with that.
I agree that it's confusing like it is right know, but I think that we should at least be consistent across languages. So, if we want to change the order of coordinates, then we should first do that in JanusGraph itself before we change it in a library like JanusGraph.Net in my opinion.

pluradj commented on src/JanusGraph.Net/IO/GraphSON/PointSerializer.cs:
```diff
@@ -0,0 +1,37 @@
+        {
+            var point = (Point) objectData;
+            return GraphSONUtil.ToTypedValue("Geoshape",
+                new Dictionary<string, object> {{"coordinates", new[] {point.Longitude, point.Latitude}}},
```
add a comment describing why the order is this way

farodin91 replied:
I would add a function to Point which returns an array with the elements in the correct order.

FlorianHockmann replied:
I created a property `Point.Coordinates` as @farodin91 suggested which has the advantage that the `PointSerializer` doesn't need to know about the order.

farodin91 commented on build.ps1:
```diff
@@ -0,0 +1,121 @@
+<#
```
Cake released a dotnet global tool with the version 0.30.0 which would allow to remove these bootstrap code and the `build.sh`.

https://cakebuild.net/blog/2018/08/cake-v0.30.0-released

FlorianHockmann replied:
Not sure if we want to do that already in the first PR. Right now, contributors who want to review this only need to install .NET Core SDK and Docker to get started.
I suppose we introduce the Cake dotnet global tool in a subsequent PR.

FlorianHockmann replied:
I created #3 for this. So, we can have a separate PR for this which probably increases the awareness of contributors for this change.

farodin91 commented on src/JanusGraph.Net/IO/GraphSON/GraphSONReaderBuilder.cs:
```diff
@@ -0,0 +1,72 @@
+    ///     Creates a <see cref="GraphSONReader" /> with the default JanusGraph deserializers and allows adding of custom
+    ///     deserializers.
+    /// </summary>
+    public class GraphSONReaderBuilder
```
The name doesn't make it clean that it contain janusgraph changes.

FlorianHockmann replied:
Ok, I'll prefix it with _Janus_, so it becomes: _JanusGraphSONReaderBuilder_. (Using _JanusGraph_ would result in _JanusGraphGraphSONReaderBuilder_ which looks odd with the duplicate _Graph_.)

farodin91 commented on src/JanusGraph.Net/IO/GraphSON/GraphSONWriterBuilder.cs:
```diff
@@ -0,0 +1,73 @@
+﻿#region License
```
The name doesn't make it clean that it contain janusgraph changes.

farodin91 commented on src/JanusGraph.Net/Text.cs:
```diff
@@ -0,0 +1,102 @@
+        /// </summary>
+        /// <param name="query">The query to search.</param>
+        /// <returns>The text predicate.</returns>
+        public static P TextContains(string query)
```
Short expression could be used also: `public static P TextContains(string query) => new P("textContains", query);`

farodin91 commented on tools/cake_version:
```diff
@@ -0,0 +1 @@
+0.29.0
```
Outdated version: 0.30.0

FlorianHockmann replied:
0.29.0 was the most recent version when I created the project, but sure, I can update to 0.30.0.

FlorianHockmann commented on test/JanusGraph.Net.IntegrationTest/RemoteConnectionFactory.cs:
```diff
@@ -0,0 +1,55 @@
+            _port = port;
+        }
+
+        public IRemoteConnection CreateRemoteConnection(string traversalSource = "gods_traversal")
```
Yes, but this factory is only used for the integration tests and there I created a graph for that graph traversal source with the Graph of the Gods for tests. That allows to use different test graphs for different purposes in the future.

FlorianHockmann commented on src/JanusGraph.Net/JanusGraph.Net.csproj:
```diff
@@ -0,0 +1,26 @@
+  <PropertyGroup>
+    <TargetFrameworks>netstandard1.3;netstandard2.0</TargetFrameworks>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
```
Yes, exactly. We don't even have that yet for Gremlin.Net (see [TINKERPOP-2018](https://issues.apache.org/jira/browse/TINKERPOP-2018)), but we should definitely do that.
I would just like to do that in a separate PR. So, we can also discuss there which tools to use and where we want to host the docs.

FlorianHockmann commented on README.md:
```diff
@@ -1,8 +1,44 @@
+Please see
+[`CONTRIBUTING.md` in JanusGraph's main repository][JanusGraph-contributing]
+for more information, including CLAs and best practices for working with
+GitHub.
```
Ok, I'll link to the main JanusGraph section on this and add a sentence about GitHub issues.

FlorianHockmann commented:
@mbrukman @pluradj @farodin91: I think that I have addressed your review comments. I also just created issues for the two aspects you mentioned where a separate PR makes sense in my opinion, namely the documentation (#2) and migrating to the Cake global tool (#3).
Could you please check whether I missed anything or whether you see other aspects that you would like to be changed?

pluradj commented on BUILDING.md:
````diff
@@ -0,0 +1,39 @@
+Other tasks can be executed by specifying them as targets:
+
+```sh
+./build.sh --target Pack
````
Once I got this command to work (see my other comment in `build.sh`), I noticed this warning:

```
home/ubuntu/.dotnet/sdk/2.2.100/Sdks/NuGet.Build.Tasks.Pack/buildCrossTargeting/NuGet.Build.Tasks.Pack.targets(202,5):
error NU5125: The 'licenseUrl' element will be deprecated. Consider using the 'license' element instead.
[/home/ubuntu/janusgraph-dotnet/src/JanusGraph.Net/JanusGraph.Net.csproj]
```

The `Pack` task completed successfully.

FlorianHockmann replied:
Yeah, I think they deprecated the `licenseUrl` element after I initially created this PR, but I'll use the new element instead.

pluradj commented on src/JanusGraph.Net/JanusGraph.Net.csproj:
```diff
@@ -0,0 +1,26 @@
+    <PackageId>JanusGraph.Net</PackageId>
+    <PackageTags>janusgraph;gremlin</PackageTags>
+    <PackageProjectUrl>http://janusgraph.org/</PackageProjectUrl>
+    <PackageLicenseUrl>https://raw.githubusercontent.com/JanusGraph/janusgraph-dotnet/master/LICENSE.txt</PackageLicenseUrl>
```
Based off this [doc](https://github.com/NuGet/Home/wiki/Packaging-License-within-the-nupkg), this could be a valid replacement to avoid the `licenseUrl` warning when the running `Pack` task:

```
<PackageLicenseExpression>Apache-2.0 AND CC-BY-4.0</PackageLicenseExpression>
```

FlorianHockmann replied:
Thanks for digging into this, Jason.
But shouldn't `Apache-2.0` be enough for the NuGet package as the package only includes code? I would rather not add the CC-BY-4.0 license here if we don't really have to.

FlorianHockmann replied:
I used just `Apache-2.0` for now, but tell me if you think that we also need to state `CC-BY-4.0` here and I'll change it accordingly.

pluradj replied:
Sounds ok. Apache should be fine since the docs aren't packaged.

FlorianHockmann commented on BUILDING.md:
```diff
@@ -0,0 +1,39 @@
+
+## Requirements
+
+* [.NET Core SDK 2.1][dotnet-sdk] is needed to
```
Good to know. I changed this requirement to make it clear that newer versions should work, too.

FlorianHockmann commented on build.sh:
```diff
@@ -0,0 +1,95 @@
+###########################################################################
+
+# Start Cake
+dotnet "$CAKE_DLL" build.cake --verbosity=$VERBOSITY --configuration=$CONFIGURATION --target=$TARGET $DRYRUN "${SCRIPT_ARGUMENTS[@]}"
```
Strange, it worked for me on the Windows Subsystem for Linux with Ubuntu, but I removed the quotes as you suggested.

FlorianHockmann commented:
@mbrukman: Could you please check whether your comments are addressed? The last commit here was in December...

FlorianHockmann commented:
@mbrukman I assume that you don't want to review this any more. If you do however intend to come back to this, then please say so. Otherwise, I'll merge this next week as I'm pretty sure that I addressed your review comments.

chupman commented:
@FlorianHockmann is this likely to be merged before Wednesday next week? Inquiring slide decks want to know.

mbrukman commented:
@FlorianHockmann — my sincere apologies for blocking this PR from being merged for so long with my old review comments which you've addressed. You're all set to merge as you have 2 approvals on this PR.

FlorianHockmann commented:
> is this likely to be merged before Wednesday next week? Inquiring slide decks want to know.

It's merged now, but I will probably make some small adjustments / updates to more recent versions before releasing a first version. We might also need a small discussion on janusgraph-dev about our release policy for such libraries as requiring 3 TSC VOTEs for every release of JanusGraph.NET, janusgraph-python, or our Docker images is too much effort in my opinion .
